test/unit/app/controllers/permissions/permissions-middleware-test.js

```diff
@@ -0,0 +1,524 @@
+import assert from 'assert'
+
+import {
+  METADATA_STORE_KEY,
+} from '../../../../../app/scripts/controllers/permissions/enums'
+
+import {
+  getPermController,
+  getEventEmitter,
+  grantPermissions,
+  getPermissionsMiddleware,
+  getApprovedPermissionsRequest,
+  rpcRequests,
+  ACCOUNT_ARRAYS,
+  CAVEATS,
+  ERRORS,
+  ORIGINS,
+  PERM_NAMES,
+  PERMS,
+} from './mocks'
+
+let permController, eventEmitter
+
+const initPermController = () => {
+  permController = getPermController()
+}
+
+const initEventEmitter = () => {
+  eventEmitter = getEventEmitter()
+}
+
+// specify an object and an key with a target object value on that object,
+// get a Promise that resolves once something is set on the target
+const getWatchPropertyPromise = (object, key) => {
```

Comment:
This is to be informed when a property is set, right? `sinon.spy` provides a simpler way to do this: https://sinonjs.org/releases/v9.0.0/spies/#using-a-spy-to-wrap-property-getter-and-setter

Reply:
All of this logic is rough, it could definitely be simplified

Reply:
`sinon.spy` didn't do the trick, but other, simpler mocks did